app/controllers/application_controller.rb

```diff
@@ -23,7 +23,13 @@ class NotAuthorizedError < StandardError
   def authorize!
     return if current_user.nil?
 
-    raise NotAuthorizedError unless Otis.config.users.include? current_user.id
+    raise NotAuthorizedError unless can?(params[:action], params[:controller])
+  end
+
+  def can?(action, resource)
+    agent = Checkpoint::Agent::Token.new('user', current_user&.id)
```

Comment:
I'm working on a new changeset, with the can? method taking an explicit user param. We don't currently have need of a guest user, except very indirectly in the `ApprovalController` (which bypasses `authorize!` in order to allow outside users to use a one-time link). I think, however, that the notion of a guest user is a little outside the scope of the ticket.

Reply:
I should restate a more specific point. One of the patterns I've really enjoyed for its clarity is passing explicit policy objects to the view. If there are multiple relevant operations you'd like to check, the policy object can hold all of the context (facts) to resolve the questions (rules). The checks are just calls to predicate methods -- the view doesn't _determine_ the context, so it should not be forced to pass it back correctly; an incidental inversion of responsibility.

For example, let's say you're in a controller action showing a single resource, and you want to condition whether to display a delete button. A policy object that is bound to the current user and that specific resource can have an explicit method like `#delete?`. This has the effect of driving the business rules out of the view and even out of the controller (whose responsibilities should mostly be HTTP and HTML).

A view helper like `can?(:action, @resource)` leaves a lot of the correctness of a feature up to the "stringly typed" view templates and implicit context (e.g., you mistakenly pass `:deleet` or the wrong resource). Calling a purpose-specific method on a collaborator (i.e., the controller constructing a policy object that has a no-arg method for the exact thing the view needs to check, like `.delete?`, and passing that in `locals`) gives you explicit testability and runtime behavior -- the policy can be tested standalone, and then the view would raise a clear exception if it called a missing method. The hard-to-test view's responsibility narrows to only calling the right method in the `if`.

The pattern of a generic `can?` in the views is convenient and reads pretty fluently, but hangs the essence of your application on runtime templates and a catchall method, both of which are highly context-dependent and nonspecific. The long-run effect is apps that are harder to test and have less deterministic behavior (aka mysterious and buggy by way of "the rules" drifting to almost anywhere in the app).

Reply:
It's not an exact match, but https://github.com/mlibrary/chipmunk/blob/master/app/controllers/v1/packages_controller.rb#L24-L33 may be a helpful example.

The `resource_policy` instance method is a consistency and convenience measure, to make sure that the right/same policy class is used throughout the controller. The `authorize!` method takes a symbol of the policy method to authorize with, `:show?`, raising unless allowed. If a view needed to check for show permission, it would receive a policy instance with the actor and package bound, and directly call the predicate `show?` method on it. (Note that `show?` is maybe not the most illustrative... checking `delete?` during from a `show` view would be more typical.)

A better would example would show this policy instance getting passed down to the view...

https://github.com/mlibrary/swapmeet/blob/master/app/controllers/listings_controller.rb#L47-L51
https://github.com/mlibrary/swapmeet/blob/master/app/views/listings/show.html.slim#L18-L22

Those are rough, but the idea is there, by way of one more indirection... A presenter combines the resource and the policy, and the net effect is the tidy and semantically sound `@listing.delete?`: a pretty direct path from the place in your template where you need to condition the display of a button to the business rule for whether that operation is permitted.

Reply:
Pushed a new batch of commits. I'll have to revisit this tomorrow. It's late enough in the day that OO wisdom is starting to look like "a steep-sided pyramid of abstractions" to quote Eric S. Raymond.